### ISSUE TITLE
No Sensor feedback (Air Purifier)

### ISSUE THREAD

**Issue by chinswain** [jghaanstra/com.xiaomi-miio#2]
The temp, humidity and air quality index do not appear to work on the device card:

![air1](https://user-images.githubusercontent.com/4403897/26993347-928c5e10-4d5a-11e7-9af7-74790215db27.PNG)

Also, they layout is too large for the mobile app:

![screenshot_20170609-190922](https://user-images.githubusercontent.com/4403897/26993351-966f0230-4d5a-11e7-92ed-9988e739607a.png)

**Comment by jghaanstra** [comment]
Thanx for reporting this.

I'll see if I can find the cause of the sensor values not updating. It's a little hard troubleshooting without the device itself. Do you have experience with installing locally through command line. I might need to add extra debug logging to tell me what is going on.

About the mobile cards, I have no control over how this is displayed. You can swipe horizontal to show the other sensor.

**Comment by chinswain** [comment]
Sure - I have it installed via CLI now so debugging is not an issue if you can add some output.
It's the Xiaomi Air Purifier 2.

I've tried scrolling across but that's not possible on the Android app.

**Comment by jghaanstra** [comment]
Ok, I have added extra debug logging in the beta branch. You can download it [here](https://github.com/jghaanstra/com.xiaomi-miio/tree/beta) and then need to run it through console (athom project --run). The Air Purifier 2 should be supported, the only reason I can think of is that the requests time out.

It's strange you cant scroll horizontally on the mobile card. I'll look into it further but it might end up as a bug report in the Athom Homey repository. I have mobile cards from other devices that indeed show 3 sensors just fine with less space between them. I also have devices with 4 sensors and that allows me to scroll horizontally.

But I'll first try to fix the sensors not showing data.

**Comment by chinswain** [comment]
Layout Fixed:

![aiq](https://user-images.githubusercontent.com/4403897/27010505-172648b8-4e9e-11e7-8f94-bafaded9762c.png)

**Comment by jghaanstra** [comment]
@chinswain would you be willing to test out the latest beta version: https://github.com/jghaanstra/com.xiaomi-miio/tree/beta

I have implemented polling for the Air Purifier and the Humidifier which should resolve this issue. Installation needs to be done through CLI and it's best if you re-pair any of these devices.

Let me know.

**Comment by chinswain** [comment]
Hi, I can't start the app:
`2/4 - Uploading to Homey @ 192.168.0.13:80...
3/4 - Running `com.xiaomi-miio`, press CTRL+C to abort...
4/4 - Debugging...

-------------------------------------------------
Driver Yeelight Bulb Color initialized ...
Driver Yeelight Bulb White initialized ...
Driver Mi Robot initialized ...
Driver Mi Air Purifier initialized ...
Driver Mi Humidifier initialized ...
ReferenceError: address is not defined
    at Object.exports.getAirPurifier.callback [as getAirPurifier] (/lib/utils.js:5:18)
    at /drivers/mi-airpurifier/driver.js:122:19
    at process.<anonymous> (/homey-app/helpers/client.js:1:1213)
    at emitTwo (events.js:87:13)
    at process.emit (events.js:172:7)
    at internal/child_process.js:729:12
    at nextTickCallbackWith0Args (node.js:489:9)
    at process._tickCallback (node.js:418:13)

--- INFO: com.xiaomi-miio has been killed ---`

**Comment by chinswain** [comment]
I can't add the air filter back to devices:
`-------------------------------------------------
Driver Yeelight Bulb Color initialized ...
Driver Yeelight Bulb White initialized ...
Driver Mi Robot initialized ...
Driver Mi Air Purifier initialized ...
Driver Mi Humidifier initialized ...
ReferenceError: address is not defined
    at Object.exports.getAirPurifier.callback [as getAirPurifier] (/lib/utils.js:5:18)
    at EventEmitter.<anonymous> (/drivers/mi-airpurifier/driver.js:22:19)
    at emitTwo (events.js:87:13)
    at EventEmitter.emit (events.js:172:7)
    at /homey-app/manager/drivers.js:268:30
    at Object.drivers.(anonymous function).ready (/homey-app/manager/drivers.js:131:9)
    at eventListeners.clientDriverPairMessage (/homey-app/manager/drivers.js:263:10)
    at /homey-app/manager/drivers.js:38:45
    at Object.Homey.ready (/homey-app/homey.js:1:1073)
    at /homey-app/manager/drivers.js:37:11

--- INFO: com.xiaomi-miio has been killed ---
`

**Comment by jghaanstra** [comment]
@chinswain sorry, there was a small type-o in a last minute change. Could you download the beta and try again.

https://github.com/jghaanstra/com.xiaomi-miio/tree/beta

**Comment by chinswain** [comment]
Getting there - I can now add it (and shows all the sensor readings in the preview) but when trying to active the filter it crashes:
`Driver Yeelight Bulb Color initialized ...
Driver Yeelight Bulb White initialized ...
Driver Mi Robot initialized ...
Driver Mi Air Purifier initialized ...
Driver Mi Humidifier initialized ...
Error: Address to device is required
    at module.exports (/node_modules/miio/lib/createDevice.js:7:30)
    at Object.module.exports.device.call.then [as device] (/node_modules/miio/lib/connectToDevice.js:6:15)
    at Object.exports.getAirPurifier.exports.getHumidifier.exports.sendCommand.callback [as sendCommand] (/lib/utils.js:54:10)
    at Object.self.capabilities.onoff.set (/drivers/mi-airpurifier/driver.js:63:23)
    at /homey-app/manager/drivers.js:197:13
    at Object.drivers.(anonymous function).ready (/homey-app/manager/drivers.js:131:9)
    at eventListeners.clientDriverMessage (/homey-app/manager/drivers.js:176:10)
    at /homey-app/manager/drivers.js:26:41
    at Object.Homey.ready (/homey-app/homey.js:1:1073)
    at /homey-app/manager/drivers.js:25:11

--- INFO: com.xiaomi-miio has been killed ---`

**Comment by jghaanstra** [comment]
Sorry, it's hard to do this without the actual device.  Could you try again, I think I fixed the above error as well now.

https://github.com/jghaanstra/com.xiaomi-miio/tree/beta

**Comment by chinswain** [comment]
Fixed! Sensor values are all there and working within flows (as tags) too.

**Comment by mhecimovic** [comment]
Dear all, I have a problem with MI air puriffier! I bought it in December and had an issue to connect it with Homey! I did everything according instructions but connection were not working! Today I noticed that system starteed to work alone after new app were released! But unfortunately not working any more! I am not sure where can be a problem! 
![6987327e-178a-415c-91eb-5e131ef5eb83](https://user-images.githubusercontent.com/37089540/52966027-86c29e80-33a6-11e9-8e15-a9fbf20195b4.png)
